### PR TITLE
Create an endpoint to view basic video info eg canvas size

### PR DESCRIPTION
**Pull request by Rosuav**
I'm not sure that this is the best way to do it, nor if it's the best place to put this code. But it's a start.

Implements #296.

### CODE REVIEW

File: src/WSRequestHandler_General.cpp [hunk 1]
```diff
@@ -181,3 +181,25 @@ HandlerResponse WSRequestHandler::HandleGetStats(WSRequestHandler* req) {
...
+ */
+HandlerResponse WSRequestHandler::HandleGetVideoInfo(WSRequestHandler* req) {
+	obs_video_info ovi; obs_get_video_info(&ovi);
+	OBSDataAutoRelease response = obs_data_create();
+	obs_data_set_int(response, "base_width", ovi.base_width);
+	obs_data_set_int(response, "base_height", ovi.base_height);
+	obs_data_set_int(response, "output_width", ovi.output_width);
+	obs_data_set_int(response, "output_height", ovi.output_height);
+	obs_data_set_double(response, "fps", (double)ovi.fps_num / ovi.fps_den); //TODO: Convert to floating-point FPS?
+	return req->SendOKResponse(response);
```
Comment (Palakis) [inline src/WSRequestHandler_General.cpp]: Would be nice to provide properties for output format, YUV range + colorspace (if YUV) and scale filter type.

File: src/WSRequestHandler_General.cpp [hunk 2]
```diff
@@ -181,3 +181,25 @@ HandlerResponse WSRequestHandler::HandleGetStats(WSRequestHandler* req) {
...
+ * @since 4.6.0 
+ */
+HandlerResponse WSRequestHandler::HandleGetVideoInfo(WSRequestHandler* req) {
+	obs_video_info ovi; obs_get_video_info(&ovi);
+	OBSDataAutoRelease response = obs_data_create();
+	obs_data_set_int(response, "base_width", ovi.base_width);
+	obs_data_set_int(response, "base_height", ovi.base_height);
+	obs_data_set_int(response, "output_width", ovi.output_width);
+	obs_data_set_int(response, "output_height", ovi.output_height);
+	obs_data_set_double(response, "fps", (double)ovi.fps_num / ovi.fps_den); //TODO: Convert to floating-point FPS?
```
[the remark this answers is not in the document]
Comment (Rosuav) [reply src/WSRequestHandler_General.cpp]: Oops, I left the TODO on that even after making the suggested change. (Previous iterations of this simply exposed fps_num and fps_den directly.) It is already exposing it as a floating-point FPS value.

**Comment by Rosuav** [comment]
Adjusted per comments/recommendations.

I'm not sure what the difference is between "Number" and "double" in the documentation. It appears that integers use "Number" but doubles use "double"? But sometimes integers use "int". The JSON data type is the same for all of them anyway.

Couldn't find a good example of rendering an enumerated value, so I went with a simple switch statement (with a fallback to integer in case new options get added to OBS).

**Comment by Palakis** [comment]
@Rosuav Thanks, gonna review this.

> Couldn't find a good example of rendering an enumerated value, so I went with a simple switch statement 

I do this as well in several places of the source code so it's good as-is.

**Comment by Palakis** [comment]
@Rosuav 
> I'm not sure what the difference is between "Number" and "double" in the documentation. It appears that integers use "Number" but doubles use "double"? But sometimes integers use "int". The JSON data type is the same for all of them anyway.

Integers should use `int` and floats/doubles use `double` in the docs, because it's the two number types supported by `obs_data_t`. Gonna remove uses of `Number` everywhere else because they're too ambiguous.

**Comment by Rosuav** [comment]
Perfect, that answers that question. Going with `int` then.

File: src/WSRequestHandler_General.cpp [hunk 3]
```diff
@@ -181,3 +181,84 @@ HandlerResponse WSRequestHandler::HandleGetStats(WSRequestHandler* req) {
...
+	obs_video_info ovi;
+	obs_get_video_info(&ovi);
+	OBSDataAutoRelease response = obs_data_create();
+	obs_data_set_int(response, "baseWidth", ovi.base_width);
+	obs_data_set_int(response, "baseHeight", ovi.base_height);
+	obs_data_set_int(response, "outputWidth", ovi.output_width);
+	obs_data_set_int(response, "outputHeight", ovi.output_height);
+	obs_data_set_double(response, "fps", (double)ovi.fps_num / ovi.fps_den);
+
+	switch (ovi.output_format) {
```
Comment (Palakis) [inline src/WSRequestHandler_General.cpp]: Can you put the four `switch` statements in separate functions that return the string value please? Makes them reusable and keeps the main function clutter-free.

File: src/WSRequestHandler_General.cpp [hunk 4]
```diff
@@ -181,3 +181,84 @@ HandlerResponse WSRequestHandler::HandleGetStats(WSRequestHandler* req) {
...
+		#define CASE(x) case x: obs_data_set_string(response, "videoFormat", #x); break;
```
Comment (Palakis) [inline src/WSRequestHandler_General.cpp]: Good idea!

File: src/WSRequestHandler_General.cpp (hunk 3 above)
Comment (Palakis) [reply src/WSRequestHandler_General.cpp]: You can put them in the same file, after include statements & defines and before member functions.

Comment (Rosuav) [reply src/WSRequestHandler_General.cpp]: Hmm. They currently don't just return a string - the fallback case has them set an integer. If they get refactored into those convenient functions, what should the fallback do? Use a static buffer to represent the integer? Return NULL and let the caller cope?

Comment (Palakis) [reply src/WSRequestHandler_General.cpp]: Use the none/unknown enum value as the fallback value. e.g `VIDEO_FORMAT_NONE`

Comment (Rosuav) [reply src/WSRequestHandler_General.cpp]: Done!

**Comment by Rosuav** [comment]
Awesome! Thank you!

**Pull request merged**